### PR TITLE
[ML-10524] Publish ML GPU containers

### PR DESCRIPTION
**Pull request by sarthfrey-db**
With the onset of GPU support for DCS, this PR aims to add 3 new containers to this repo. As per the README:

```
These images attempt to mimic the Databricks Machine Learning Runtime workflows that use GPUs.
`ml-gpu-base` extends `standard` with MLR python libraries, without support for ganglia or
non-python libraries such as hive libraries or jvm libraries. `ml-gpu-cuda` installs CUDA 10.1
onto `ml-gpu-base`. `ml-gpu-rapids` installs [RAPIDS](https://rapids.ai/) onto `ml-gpu-cuda`.
```

### CODE REVIEW

File: ubuntu/ml/gpu/cuda/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,57 @@
+FROM sarthfrey/ml-gpu-base:latest
+
+# Taken from https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/ubuntu16.04/10.1/base/Dockerfile
```
Comment (sarthfrey-db) [inline ubuntu/ml/gpu/cuda/Dockerfile:3]: had issues with exposing `libcuda.so` with `--gpus all` when this image was built with our internal databricks cuda installation workflow. Copying and using NVIDIA's cuda image installation steps solved that problem.

File: ubuntu/ml/gpu/rapids/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
+FROM sarthfrey/ml-gpu-cuda:latest
+
+RUN /databricks/conda/bin/conda install -n ${DATABRICKS_ROOT_CONDA_ENV} \
+    -c rapidsai -c nvidia -c conda-forge -c defaults \
+    rapids=0.13 python=3.7 cudatoolkit=10.1
```
Comment (sarthfrey-db) [inline ubuntu/ml/gpu/rapids/Dockerfile:5]: This does not work

Comment (mengxr) [reply ubuntu/ml/gpu/rapids/Dockerfile:5]: Have a minimal yml file that describes the rapids.ai dependencies as well as databricks minimal dependencies (ipython, pandas, pyarrow, etc). Then use that yaml to create a new env, which is simpler for conda to resolve dependencies.

Install the latest miniconda (4.8.3), which is faster and better at solving.

**Comment by mengxr** [comment]
GPU containers:
1. gpu-minimal (FROM databricks:minimal, only installs nvidia CUDA and GPU libraries via ubuntu)
2. gpu-python (FROM databricks:gpu-minimal, only installs the latest miniconda and create an environment call "gpu" with minimal dependencies for Spark (do not include cudatoolkit from conda)
3. gpu-rapidsai (FROM databricks:gpu-python, add rapids.ai to the "gpu" env with conda env update)
Optionally:
4. gpu-pytorch
5. gpu-tensorflow-2

File: ubuntu/ml/gpu/cuda/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ENV NVIDIA_REQUIRE_CUDA "cuda>=10.1 brand=tesla,driver>=384,driver<385 brand=tesla,driver>=396,driver<397 brand=tesla,driver>=410,driver<411"
+
+# Install CUDA, NCCL, and CUDNN
+RUN wget https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64/nvidia-machine-learning-repo-ubuntu1804_1.0.0-1_amd64.deb \
+    && dpkg -i nvidia-machine-learning-repo-ubuntu1804_1.0.0-1_amd64.deb \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends --allow-downgrades \
+      libnccl2=2.4.8-1+cuda10.1 \
+      libnccl-dev=2.4.8-1+cuda10.1 \
+      cuda-libraries-10-1 \
```
Comment (mengxr) [inline ubuntu/ml/gpu/cuda/Dockerfile:47]: It might be simpler to get the cuda-toolkit-10-1 directly.

File: ubuntu/ml/gpu/cuda/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      libcudnn7=7.6.4.38-1+cuda10.1 \
+      libcudnn7-dev=7.6.4.38-1+cuda10.1 \
+      libcublas10=10.2.1.243-1 \
```
Comment (mengxr) [inline ubuntu/ml/gpu/cuda/Dockerfile:50]: leave a comment about this tricky dependency

File: ubuntu/gpu/python/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,30 @@
+FROM sarthfrey/gpu-cuda:10.1
```
Comment (evanye) [inline ubuntu/gpu/python/Dockerfile:1]: we'll have to fix this before merging. When you're ready to build the images I can push them to databricksruntime public docker hub and you can update these.

File: ubuntu/gpu/rapids/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,5 @@
+FROM sarthfrey/gpu-rapids:latest
```
Comment (evanye) [inline ubuntu/gpu/rapids/Dockerfile:1]: same here

File: ubuntu/gpu/minimal/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,54 @@
+FROM databricksruntime/minimal:latest
+
+# Taken from https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/ubuntu16.04/10.1/base/Dockerfile
```
Comment (mengxr) [inline ubuntu/gpu/minimal/Dockerfile:3]: Note: BSD license.

File: ubuntu/gpu/README.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# GPU Containers
+
+These images are for common use cases that relate to GPUs.
+`gpu-minimal` extends [`minimal`](https://github.com/databricks/containers/tree/master/ubuntu/minimal) with CUDA, without support for python, ganglia, or
+non-python libraries such as hive libraries or jvm libraries. `gpu-python` also installs the minimum python dependencies and will be a common image to start with for many python Databricks GPU users. `gpu-rapids` installs [RAPIDS](https://rapids.ai/) onto the `gpu-python` image. `gpu-rapids` installs TensorFlow onto the `gpu-python` image. `gpu-rapids` installs PyTorch onto the `gpu-python` image.
```
Comment (mengxr) [inline ubuntu/gpu/README.md:5]: * gpu-rapids -> gpu-tensorflow
* gpu-rapids -> gpu-pytorch

File: ubuntu/gpu/minimal/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    && apt-get update \
+    && apt-get install -y --no-install-recommends --allow-downgrades \
+      cuda-toolkit-10-1 \
+      libnccl2=2.4.8-1+cuda10.1 \
+      libnccl-dev=2.4.8-1+cuda10.1 \
+      libcudnn7=7.6.4.38-1+cuda10.1 \
+      libcudnn7-dev=7.6.4.38-1+cuda10.1 \
+      libcublas10=10.2.1.243-1 \
+      libcublas-dev=10.2.1.243-1 \
+    && apt-get clean
```
Comment (mengxr) [inline ubuntu/gpu/minimal/Dockerfile:54]: minor: clean is not needed

File: ubuntu/gpu/python/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    # Conda must be installed at /databricks/conda
+    && /bin/bash miniconda.sh -b -p /databricks/conda \
+    && rm miniconda.sh \
+    && /databricks/conda/bin/conda install --name base conda=4.8.3
+
+COPY env.yml /databricks/.conda-env-def/env.yml
+
+RUN /databricks/conda/bin/conda env create --file /databricks/.conda-env-def/env.yml \
+    # Source conda.sh for all login shells.
+    && ln -s /databricks/conda/etc/profile.d/conda.sh /etc/profile.d/conda.sh
```
Comment (mengxr) [inline ubuntu/gpu/python/Dockerfile:21]: minor: this line should go to the conda RUN block

**Comment by sarthfrey** [comment]
@mengxr @evanye I believe this wasn't tracked in a ticket so letting you know in case this goes untracked :)

File: ubuntu/gpu/python/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    && (echo 'e5e5b4cd2a918e0e96b395534222773f7241dc59d776db1b9f7fedfcb489157a  miniconda.sh' | shasum -a 256 -c) \
+    # Conda must be installed at /databricks/conda
+    && /bin/bash miniconda.sh -b -p /databricks/conda \
+    && rm miniconda.sh \
+    && /databricks/conda/bin/conda install --name base conda=4.8.3
+
+COPY env.yml /databricks/.conda-env-def/env.yml
+
+# Create conda environment
+RUN /databricks/conda/bin/conda env create --file /databricks/.conda-env-def/env.yml \
```
Comment (brookskd) [inline ubuntu/gpu/python/Dockerfile:20]: I had to remove the trailing \ from this line to get this to build, I think it's a typo.

Comment (mengxr) [reply ubuntu/gpu/python/Dockerfile:20]: Fixed in the new PR: #24.

**Pull request closed without merging**